bdns/problem.go

```diff
@@ -9,30 +9,55 @@ import (
 	"fmt"
 	"net"
 
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/miekg/dns"
 	"github.com/letsencrypt/boulder/probs"
 )
 
-const detailDNSTimeout = "DNS query timed out"
-const detailDNSNetFailure = "DNS networking error"
-const detailServerFailure = "Server failure at resolver"
+type dnsError struct {
+	recordType uint16
+	hostname   string
+	// Exactly one of rCode or underlying should be set.
+	underlying error
+	rCode      int
+}
+
+func (d dnsError) Error() string {
+	var detail string
+	if d.underlying != nil {
+		if netErr, ok := d.underlying.(*net.OpError); ok {
+			if netErr.Timeout() {
+				detail = detailDNSTimeout
+			} else {
+				detail = detailDNSNetFailure
+			}
+		}
+	} else if d.rCode != dns.RcodeSuccess {
+		detail = dns.RcodeToString[d.rCode]
+	} else {
+		detail = detailServerFailure
+	}
+	return fmt.Sprintf("DNS problem: %s looking up %s for %s", detail,
+		dns.TypeToString[d.recordType], d.hostname)
+}
+
+const detailDNSTimeout = "query timed out"
+const detailDNSNetFailure = "networking error"
+const detailServerFailure = "server failure at resolver"
 
 // ProblemDetailsFromDNSError checks the error returned from Lookup...  methods
 // and tests if the error was an underlying net.OpError or an error caused by
 // resolver returning SERVFAIL or other invalid Rcodes and returns the relevant
 // core.ProblemDetails. The detail string will contain a mention of the DNS
 // record type and domain given.
-func ProblemDetailsFromDNSError(recordType, domain string, err error) *probs.ProblemDetails {
-	detail := detailServerFailure
-	if netErr, ok := err.(*net.OpError); ok {
-		if netErr.Timeout() {
-			detail = detailDNSTimeout
-		} else {
-			detail = detailDNSNetFailure
+func ProblemDetailsFromDNSError(err error) *probs.ProblemDetails {
```

Comment:
So, I'm thinking more and I'm thinking this helper isn't helping as much now that we need to handle DeadlineExceeded and Canceled. Since the deadline could be missed not because of the query, per se (maybe a really long GC or something), it doesn't seem right for them to be referred to as DNS errors by this helper.

Hrm.